service/history/workflow/update/update.go

```diff
@@ -159,31 +163,86 @@ func newCompleted(
 // WaitOutcome observes this Update's completion, returning the Outcome when it
 // is available. This call will block until the Outcome is known or the provided
 // context.Context expires. It is safe to call this method outside of workflow lock.
-func (u *Update) WaitOutcome(ctx context.Context) (*updatepb.Outcome, error) {
-	return u.outcome.Get(ctx)
+func (u *Update) WaitOutcome(ctx context.Context) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error) {
+	outcome, err := u.outcome.Get(ctx)
+	if err != nil {
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, outcome, err
+	}
+	return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED, outcome, nil
 }
 
 // WaitAccepted blocks on the acceptance of this update, returning nil if has
 // been accepted but not yet completed or the overall Outcome if the update has
 // been completed (including completed by rejection). This call will block until
 // the acceptance occurs or the provided context.Context expires.
 // It is safe to call this method outside of workflow lock.
-func (u *Update) WaitAccepted(ctx context.Context) (*updatepb.Outcome, error) {
+func (u *Update) WaitAccepted(ctx context.Context) (enumspb.UpdateWorkflowExecutionLifecycleStage, *updatepb.Outcome, error) {
 	if u.outcome.Ready() {
-		// being complete implies being accepted, return the completed outcome
+		// Being complete implies being accepted; return the completed outcome
 		// here because we can.
-		return u.outcome.Get(ctx)
+		return u.WaitOutcome(ctx)
 	}
 	fail, err := u.accepted.Get(ctx)
 	if err != nil {
-		return nil, err
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, nil, err
 	}
 	if fail != nil {
-		return &updatepb.Outcome{
+		outcome := &updatepb.Outcome{
 			Value: &updatepb.Outcome_Failure{Failure: fail},
-		}, nil
+		}
+		return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED, outcome, nil
+	}
+	return enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ACCEPTED, nil, nil
+}
+
+// WaitLifecycleStage waits until the Update has reached at least `waitStage` or
+// a timeout. If the Update reaches `waitStage` with no timeout, the most
+// advanced stage known to have been reached is returned, along with the outcome
+// if any. If there is a timeout due to the supplied soft timeout, then
+// unspecified stage and nil outcome are returned, without an error. If there is
+// a timeout due to context deadline expiry, then the error is returned as usual.
+func (u *Update) WaitLifecycleStage(ctx context.Context, waitStage enumspb.UpdateWorkflowExecutionLifecycleStage, softTimeout time.Duration) (stage enumspb.UpdateWorkflowExecutionLifecycleStage, outcome *updatepb.Outcome, err error) {
```

Comment:
It seems like Lifecycle Stage returned by this function is defined by the Lifecycle Stage received and the outcome of the WaitXXX function. Am I right?
Assuming I am right, I would keep all the code as it was creating a new function which can produce new Lifecycle Stage given old Lifecycle Stage and the produced Outcome. This way the table you presented will be encoded as a single function, so you would not need a separate document to explain the behavior.